cmd/opm/root/add.go

```diff
@@ -0,0 +1,113 @@
+)
+
+type add struct {
+	logger     *logrus.Entry
+	configsDir string
+	debug      bool
+	caFile     string
+	pullTool   string
+	skipTLS    bool
+}
+
+func newAddCmd() *cobra.Command {
+	logger := logrus.New()
+	a := add{
+		logger:   logrus.NewEntry(logger),
+		pullTool: "none",
+	}
+	rootCmd := &cobra.Command{
+		Use:   "add <configs_path> <bundle_image1> <bundle_image2>........<bundle_imageN>",
+		Short: "add operator bundle/s to a catalog of packages",
+		Long:  `add operator bundles to a directory of configs representing packages in the catalog`,
+		Args:  cobra.MinimumNArgs(2),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			a.configsDir = args[0]
+			if a.debug {
+				logger.SetLevel(logrus.DebugLevel)
+			}
+			return nil
+		},
+
+		RunE: a.addFunc,
+	}
+
+	rootCmd.Flags().BoolVar(&a.debug, "debug", false, "enable debug logging")
+	rootCmd.Flags().StringVarP(&a.caFile, "ca-file", "", "", "the root Certificates to use with this command")
+	rootCmd.Flags().BoolVar(&a.skipTLS, "skip-tls", false, "disable TLS verification")
+	return rootCmd
+}
+
+func (a *add) addFunc(cmd *cobra.Command, args []string) error {
+	rootCAs, err := certs.RootCAs(a.caFile)
+	if err != nil {
+		return fmt.Errorf("failed to get RootCAs: %v", err)
+	}
+	reg, err := containerdregistry.NewRegistry(containerdregistry.SkipTLS(a.skipTLS), containerdregistry.WithLog(a.logger), containerdregistry.WithRootCAs(rootCAs))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := reg.Destroy(); err != nil {
+			a.logger.Errorf("error destroying local cache: %v", err)
+		}
+	}()
+	bundles := []action.InputBundle{}
+	for _, ref := range args[1:] {
+		simpleRef := image.SimpleReference(ref)
+		tmpDir, err := ioutil.TempDir("./", "bundle_tmp")
+		if err != nil {
+			return fmt.Errorf("error creating temp directory to unpack bundle image %q in:%v", simpleRef.String(), err)
+		}
+		defer func() {
+			a.logger.Infof("Removing temp directory %q bundle was unpacked in", tmpDir)
+			if err := os.RemoveAll(tmpDir); err != nil {
+				a.logger.Errorf("error removing temp directory %q bundle was unpacked in: %v", tmpDir, err)
+			}
+		}()
+		err = wait.PollImmediate(retryInterval, timeout, func() (done bool, e error) {
+			a.logger.Infof("Pulling bundle %q", simpleRef.String())
+			e = reg.Pull(cmd.Context(), simpleRef)
+			if err != nil {
+				a.logger.Info("Error pulling bundle %q. Trying to pull again.", simpleRef.String())
+				return false, err
+			}
+			return true, nil
+		})
+		if err != nil {
+			return fmt.Errorf("error pulling image %q into registry:%v", simpleRef.String(), err)
+		}
+		a.logger.Infof("Unpacking bundle %q into %q", simpleRef.String(), tmpDir)
+		err = reg.Unpack(cmd.Context(), simpleRef, tmpDir)
+		if err != nil {
+			return fmt.Errorf("error unpacking image %q: %v", simpleRef.String(), err)
+		}
+		bundles = append(bundles, action.InputBundle{Dir: tmpDir, ImgRef: simpleRef})
+	}
+	request := action.AddConfigRequest{
+		Bundles:    bundles,
```

Comment:
IMO, all the logic to extract the bundles should be encapsulated in the `AddConfigRequest.AddToConfig()` function somewhere. Otherwise other theoretical clients of this package would have to do all this work themselves.

So I think we should just pass `args[1:]` into the AddConfigRequest as its Bundles:
```suggestion
		Bundles:    args[1:],
```

Reply:
@joelanford I noticed that's what some of the current interfaces in the `opm index` land does that, i.e bakes in the work to extract and unpack bundles insides their implementations. However, that means that each of those interface implementations instantiate their own `image.Registry`, and @njhale pointed out earlier that that's what makes makes writing e2e tests for these hard.  

>Otherwise other theoretical clients of this package would have to do all this work themselves.

What's the con of telling the client "unpack the bundles yourself they way you desire, and then give me the unpacked bundles"? 

I kind of see the point that that separation of concern from `AddToConfig` and similar interfaces makes them easily testable.

Reply:
If that's the only problem, then I'd suggest making the image.Registry a field on the `AddConfigRequest` struct.

Reply:
> What's the con of telling the client "unpack the bundles yourself they way you desire, and then give me the unpacked bundles"?

Mainly it's that that callers of this function probably don't really care about the details of image unpacking. They're thinking "here's my image, add it to my declarative config".

I think having the image registry have an automatic default, but making it overridable is the way to go. We could use the functional options pattern to make this clean: https://medium.com/@dreissenzahn/functional-options-in-go-35279f535c6

Reply:
How does it tie in with what we've been talking about in terms of not baking in external tools to deal with images? 
If we change `BundleAdder` to accept an image instead of a directory, does it make it harder for us to port to `opm index add <bundle-directory> <configs-directory>`?